### PR TITLE
fix: Phone users to abide by guest policy

### PR DESCRIPTION
**Pull request by zhem0004**
Closes:#9531 and #4787
<h3>What does this PR do?</h3>
It allows moderator to have control over phone user ability to join the meeting

<h3>Motivation</h3>
Previously: A phone user can join the meeting whenever they like
Currently: Phone user's ability to join the meeting depends on guest policy set by moderator

<h3>Notes</h3>
<ul>
<li>If moderator sends a message to waiting room, dial-in users cannot see it as it is a phone call.</li>
<li>If moderator ends a meeting while phone user is in waiting room all ends</li>
<li>If Guest Policy is to deny all, phone call ends.</li>
</ul>

This is an example of how it would look( bohdan is a dial-in user)

Screenshot from 2022-01-19 14-22-45

<h3>Explanation on second issue close</h3>
The cause of #4787 was that phone users were not registered in Users2x, since registering them was necessary for #9531 , the PR ended up closing both.

You can see here ( https://github.com/bigbluebutton/bigbluebutton/blob/develop/akka-bbb-apps/src/main/scala/org/bigbluebutton/core2/message/handlers/MuteMeetingCmdMsgHdlr.scala ) that 'Mute all' was functioning properly as it iterated over VoiceUsers. While LockSettings iterated over Users2x ( https://github.com/bigbluebutton/bigbluebutton/blob/develop/akka-bbb-apps/src/main/scala/org/bigbluebutton/LockSettingsUtil.scala ).

### CODE REVIEW

File: akka-bbb-apps/src/main/scala/org/bigbluebutton/core/apps/voice/UserJoinedVoiceConfEvtMsgHdlr.scala [hunk 1]
```diff
@@ -39,5 +36,48 @@ trait UserJoinedVoiceConfEvtMsgHdlr extends SystemConfiguration {
...
+        val event = MsgBuilder.buildGuestsWaitingForApprovalEvtMsg(meetingId, m.intId, guests)
+        outGW.send(event)
+      }
+      // Meteor should only listen for this single message
+      val event = MsgBuilder.buildGuestsWaitingForApprovalEvtMsg(meetingId, "nodeJSapp", guests)
+      outGW.send(event)
+    }
+
+    def denyUser() = {
+      val g = GuestApprovedVO(msg.body.voiceUserId, GuestStatus.DENY)
```
Comment (antobinary) [inline akka-bbb-apps/src/main/scala/org/bigbluebutton/core/apps/voice/UserJoinedVoiceConfEvtMsgHdlr.scala:52]: Please rename `g` to something more explicit

File: akka-bbb-apps/src/main/scala/org/bigbluebutton/core/apps/voice/UserJoinedVoiceConfEvtMsgHdlr.scala [hunk 2]
```diff
@@ -39,5 +36,48 @@ trait UserJoinedVoiceConfEvtMsgHdlr extends SystemConfiguration {
         "freeswitch"
       )
     }
+
+    def notifyModeratorsOfGuestWaiting(guests: Vector[GuestWaiting], users: Users2x, meetingId: String): Unit = {
+      val mods = Users2x.findAll(users).filter(p => p.role == Roles.MODERATOR_ROLE && p.clientType == ClientType.FLASH)
```
Comment (antobinary) [inline akka-bbb-apps/src/main/scala/org/bigbluebutton/core/apps/voice/UserJoinedVoiceConfEvtMsgHdlr.scala:41]: `ClientType.FLASH` is obsolete, were you trying to filter for web users?

File: akka-bbb-apps/src/main/scala/org/bigbluebutton/core/apps/voice/UserJoinedVoiceConfEvtMsgHdlr.scala [hunk 3]
```diff
@@ -39,5 +36,48 @@ trait UserJoinedVoiceConfEvtMsgHdlr extends SystemConfiguration {
...
       )
     }
+
+    def notifyModeratorsOfGuestWaiting(guests: Vector[GuestWaiting], users: Users2x, meetingId: String): Unit = {
+      val mods = Users2x.findAll(users).filter(p => p.role == Roles.MODERATOR_ROLE && p.clientType == ClientType.FLASH)
+      mods foreach { m =>
+        val event = MsgBuilder.buildGuestsWaitingForApprovalEvtMsg(meetingId, m.intId, guests)
+        outGW.send(event)
+      }
+      // Meteor should only listen for this single message
```
Comment (antobinary) [inline akka-bbb-apps/src/main/scala/org/bigbluebutton/core/apps/voice/UserJoinedVoiceConfEvtMsgHdlr.scala:46]:
```suggestion
      // bbb-html should only listen for this single message

```

File: akka-bbb-apps/src/main/scala/org/bigbluebutton/core/apps/voice/UserJoinedVoiceConfEvtMsgHdlr.scala [hunk 4]
```diff
@@ -5,6 +5,10 @@ import org.bigbluebutton.common2.msgs._
 import org.bigbluebutton.core.models.VoiceUsers
 import org.bigbluebutton.core.running.{ LiveMeeting, MeetingActor, OutMsgRouter }
 import org.bigbluebutton.core2.message.senders.MsgBuilder
+import org.bigbluebutton.core.models._
+import org.bigbluebutton.core2.message.senders.MsgBuilder
+import org.bigbluebutton.core.apps.users.UsersApp
+import org.bigbluebutton.core.models.UserLeftFlag
```
Comment (antobinary) [inline akka-bbb-apps/src/main/scala/org/bigbluebutton/core/apps/voice/UserJoinedVoiceConfEvtMsgHdlr.scala:11]: already imported above via `models._`

Comment (antobinary) [reply akka-bbb-apps/src/main/scala/org/bigbluebutton/core/apps/voice/UserJoinedVoiceConfEvtMsgHdlr.scala:11]: Same for `import org.bigbluebutton.core.models.VoiceUsers` - either keep the explicit imports or go back and remove the ones which are now imported twice

File: akka-bbb-apps/src/main/scala/org/bigbluebutton/core/apps/voice/UserJoinedVoiceConfEvtMsgHdlr.scala [hunk 5]
```diff
@@ -13,17 +17,10 @@ trait UserJoinedVoiceConfEvtMsgHdlr extends SystemConfiguration {
...
-      log.info("Ejecting banned voice user " + msg)
-      val event = MsgBuilder.buildEjectUserFromVoiceConfSysMsg(
-        props.meetingProp.intId,
-        props.voiceProp.voiceConf,
-        msg.body.voiceUserId
-      )
-      outGW.send(event)
-    } else {
+    val guestPolicy = GuestsWaiting.getGuestPolicy(liveMeeting.guestsWaiting)
+    val currentTime = System.currentTimeMillis()
```
Comment (antobinary) [inline akka-bbb-apps/src/main/scala/org/bigbluebutton/core/apps/voice/UserJoinedVoiceConfEvtMsgHdlr.scala:22]: You are only using currentTime once and actually much later. Please consider calling System.... directly there or define ~ in the same block

**Review by Tainan404 — APPROVED** [review]

**Pull request merged**